bson/src/main/org/bson/codecs/pojo/ConventionAnnotationImpl.java

```diff
@@ -100,6 +111,14 @@ private void processPropertyAnnotations(final ClassModelBuilder<?> classModelBui
         }
     }
 
+    private <T> boolean  bsonRepIsValid(final BsonType bsonRep, final Class<T> clazz) {
```

Comment:
Even though we're throwing exceptions in the `Codec` constructor now, I decided to leave this check in there by the principle of catching errors as soon as we can. Is that good or unnecessary?

Reply:
Can users write their own RepresentationConfigurable codecs? If so can we validate them here?

Reply:
@bdeleonardis1 Have you thought about what to do here?

Reply:
Sorry @rozza I missed this question the first time. If users can write their own `RepresentationConfigurable` codecs I don't think we can check for validity here because there is no way of predicting what types of conversions these codecs will allow and we don't have even have access to the Codec yet at this point in the execution. I think the best option is to just let codecs do error checking as they'red constructed (i.e. in the call to withRepresentation). If they're provided with an invalid BsonRep in that method call, they can throw an exception before ever being created.

Reply:
Users often may want to do their own thing and with representations you've done all the groundwork to make it extensible. 

I'd remove this check here and add it later when fetching the codec.  You may also want to document how to create a custom BsonRepresentation (@jyemin do you think its worth it?)

Reply:
@rozza I think I'm a little confused. I thought we were just trying to allow users to to create Codecs that implement `RepresentationConfigurable` so that they can create any conversions that they want.

Or are you saying we want to allow them to create a custom BsonRepresentation that just works with our built in Codecs (for example `@BsonRepresentation(BsonType.INT32)` on a String and then our `StringCodec` handles it)? 

If it's the first one, I don't see why we can't check for validity in `withRepresentation`. Our built in codecs would know all of the conversions that they support, so they can just check them on construction. That would prevent us from having to do type checking/casting where you highlighted in the code below.

If it's the second one, that seems a lot harder. We would need to add support to each of our codecs for all possible conversions, right?

Reply:
Sorry, I wasn't very clear.  I can think of two main error scenarios:

1) If a user added `@BsonRepresentation(BsonType.INT32)` and the codec is `RepresentationConfigurable` then the codec should validate the BsonType.
2) If a user adds `@BsonRepresentation(BsonType.STRING)` and the resulting codec isn't `RepresentationConfigurable` then the PojoCodec should error.

Reply:
Ah I see what you're saying now. So I'll have checks in `withRepresentation` that will handle your case 1, and an additional case in the place that you highlighted that will handle your case 2. Thanks for clarifying for me.

Reply:
Done. I already had test cases for each of these cases in `PojoCustomTest.java` so I didn't add any cases. I did change the exception being thrown (and expected in the tests) to `IllegalArgumentException` though because `CodecConfigurationException` gets caught in `getCodecFromPropertyRegistry` and we want in thrown.